### ISSUE TITLE
getDisabledIcon() doesn't work as expected

### ISSUE THREAD

**Issue by frankb33** [violetlib/vaqua#15]
if I want to have a disabled icon I usually call getDisabledIcon(). If I used that with VAqua (tested V9 and V10) I get an unchanged  icon. For all other L&F I get a visual different one. Here is a minimal code to reproduce it.

import javax.swing.*;

public class TestDisabledIcon
{
    public static void main(String[] args)
        throws Exception
    {
//        UIManager.setLookAndFeel("org.violetlib.aqua.AquaLookAndFeel");
//        UIManager.setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel");
        UIManager.setLookAndFeel("com.apple.laf.AquaLookAndFeel");

        JFrame jf = new JFrame("test");
        jf.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        ImageIcon ii = new ImageIcon("checker_red.png");  // any will do 
        JButton b1 = new JButton(ii);
        JButton b2 = new JButton(b1.getDisabledIcon());

        JPanel p = new JPanel();
        p.add(b1);
        p.add(b2);
        jf.getContentPane().add(p);
        jf.pack();
        jf.setVisible(true);
    }
}

I would highly appreciate if the fix would not only in V10 but also in V9 because I need the latter for my next release.

**Comment by cbfiddle** [comment]
> On Nov 6, 2021, at 9:43 AM, Frank Berger ***@***.***> wrote:
> 
>     JButton b2 = new JButton(b1.getDisabledIcon());
> 

I’m not sure why you would want to install a disabled icon in a button.

I think the general idea is that the disabled icon is used when the button is disabled; if you want to control the button state you set the button or its model to the disabled state.

  Alan

**Comment by frankb33** [comment]
This is only demo code to show the issue. In my real program I have an icon derived from a larger image on the fly and need a disabled one.

**Comment by cbfiddle** [comment]
I am still not understanding the situation.

Do you want the LAF to display a disabled version, or is there some reason why you want the LAF to create a disabled version and return it to your application?

**Comment by frankb33** [comment]
I have a JButton with only one icon.  When I disable the JButton and I haven't provided a disabled icon, a disabled icon is generated by the L&F, usually a grey one. With VAqua the icon isn't changed so there is no visible feedback for the user that the Button is disabled. in a nutshel: getDisabledIcon() returns a grey icon for all other L&Fs and the same icon for VAqua.
I hope I could explain the issue.

**Comment by cbfiddle** [comment]
Actually, that is helpful.

Unlike other LAFs, VAqua creates a disabled icon as needed when the button is painted in the disabled state. That icon is not returned by getDisabledIcon().

If getDisabledIcon() returns an icon that is not an instance of UIResource, then VAqua uses it directly, which sounds like it might be the problem.

When you say that getDisabledIcon() returns the same icon, is it a copy that is a UIResource or your original icon?

If it is the original icon, then setDisabledIcon(null) might fix the problem.

Otherwise, I guess I would need to see a test case that displays incorrectly.

  Alan

**Comment by frankb33** [comment]
The exact use case is as following: I have a JButton that has two states, that I use as a kind of ToggleButton (I do some customizations in a kind of JButton for toolbars and the like, so it behaves in several OS/L&F as I need it). So I had the choice of using the ToggleButton and duplicate the code or use a JButton and a a few lines for switching between the states. I have chose the latter. For the 2nd state I need the generated icon that looks like the disabled button.
I fully understand that there is a trade-off between being as MacOS like as possible and conformant to other L&F, but for a cross platform program that is not trival it hurts WORA, so I would appreciate if VAqua returns the disabled icon.

I refined the example:
public class TestDisabledIcon
{
    public static void main(String[] args)
        throws Exception
    {
        UIManager.setLookAndFeel("org.violetlib.aqua.AquaLookAndFeel");
//        UIManager.setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel");
//        UIManager.setLookAndFeel("com.apple.laf.AquaLookAndFeel");


        JFrame jf = new JFrame("test");
        jf.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        ImageIcon  ii = new ImageIcon("checker_red.png");  // any will do
        JButton b1 = new JButton(ii);
        JButton b2 = new JButton(b1.getDisabledIcon());
        JButton b3 = new JButton(ii);

        JPanel p = new JPanel();
        p.add(b1);
        p.add(b2);
        p.add(b3);
        b2.setEnabled(false);
        b2.setEnabled(false);
        jf.getContentPane().add(p);
        jf.pack();
        jf.setVisible(true);

The  left button is  enabled; the middle has the generated disabled icon. I would like to get that icon to use it, but I don't get it as shown in Button3 . The API says that a L&F might ignore this, but VAqua is the first I experienced this behavior
